### PR TITLE
Refactor backtrace creation to fix disabled backtrace handling.

### PR DESCRIPTION
**Pull request by aardvark179**
Refactors our creation of ruby exceptions in Java to ensure we don't miss any cases where we should omit filling in the backtrace, and to separate out the logic for non-exception filling cases so we don't accidental omit it when we shouldn't.

### CODE REVIEW

File: src/main/java/org/truffleruby/core/thread/ThreadNodes.java [hunk 1]
```diff
@@ -573,7 +573,8 @@ public static void raiseInThread(RubyContext context, DynamicObject rubyThread,
 
             context.getSafepointManager().pauseRubyThreadAndExecute(rubyThread, currentNode, (currentThread, currentNode1) -> {
                 if (Layouts.EXCEPTION.getBacktrace(exception) == null) {
-                    Backtrace backtrace = context.getCallStack().getBacktrace(currentNode1);
+                    DynamicObject exceptionClass = Layouts.BASIC_OBJECT.getLogicalClass(exception);
+                    Backtrace backtrace = context.getCallStack().getBacktraceForExceptoin(currentNode1, exceptionClass);
```
Comment (bjfish) [inline src/main/java/org/truffleruby/core/thread/ThreadNodes.java]: Typ0 `Exceptoin`?

File: src/main/java/org/truffleruby/platform/posix/TrufflePosixHandler.java [hunk 2]
```diff
@@ -28,16 +28,10 @@ public TrufflePosixHandler(RubyContext context) {
...
     @Override
     public void error(Errno errno, String methodName) {
         // TODO CS 17-Apr-15 - not specialised, no way to build a good stacktrace, missing content for error messages
-
-        throw new RaiseException(ExceptionOperations.createSystemCallError(
-                context.getCoreLibrary().getErrnoClass(errno),
-                context.getCoreLibrary().getNil(),
-                context.getCallStack().getBacktrace(null),
-                errno.intValue()));
+        throw new RaiseException(ExceptionOperations.createSystemCallError(context, errno, context.getCoreLibrary().getNil()));
```
Comment (bjfish) [inline src/main/java/org/truffleruby/platform/posix/TrufflePosixHandler.java]: There's also `context.getCoreExceptions().errnoError(` or `systemCallError(`

File: src/main/java/org/truffleruby/core/exception/ExceptionOperations.java [hunk 3]
```diff
@@ -63,18 +67,26 @@ public static DynamicObject backtraceAsRubyStringArray(RubyContext context, Dyna
...
         return Layouts.SYSTEM_CALL_ERROR.createSystemCallError(Layouts.CLASS.getInstanceFactory(rubyClass), message, backtrace, errno);
     }
 
+    // because the factory is not constant
+    @TruffleBoundary
+    public static DynamicObject createSystemCallError(RubyContext context, Errno errno, Object message) {
+        DynamicObject rubyClass = context.getCoreLibrary().getErrnoClass(errno);
+        Backtrace backtrace = context.getCallStack().getBacktraceForException(null, rubyClass, null);
+        context.getCoreExceptions().showExceptionIfDebug(rubyClass, message, backtrace);
+        return Layouts.SYSTEM_CALL_ERROR.createSystemCallError(Layouts.CLASS.getInstanceFactory(rubyClass), message, backtrace, errno.intValue());
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/exception/ExceptionOperations.java]: Could this call the method above?

File: src/main/java/org/truffleruby/core/string/ConvertBytes.java [hunk 4]
```diff
@@ -288,7 +288,7 @@ private long stringToLong(int nptr, int[] endptr, int base) {
         return 0;
     }
 
-    @TruffleBoundary
+    @TruffleBoundary(transferToInterpreterOnException = false)
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/string/ConvertBytes.java]: That's an unrelated change, would be to good to split it in another commit at least.

File: src/main/java/org/truffleruby/extra/TrufflePosixNodes.java [hunk 5]
```diff
@@ -377,7 +377,7 @@ public int utimes(DynamicObject path, DynamicObject pointer) {
     @CoreMethod(names = "mkdir", isModuleFunction = true, required = 2, lowerFixnum = 2)
     public abstract static class MkdirNode extends CoreMethodArrayArgumentsNode {
 
-        @TruffleBoundary
+        @TruffleBoundary(transferToInterpreterOnException = false)
```
Comment (eregon) [inline src/main/java/org/truffleruby/extra/TrufflePosixNodes.java]: (same here)

File: src/main/java/org/truffleruby/language/CallStackManager.java [hunk 6]
```diff
@@ -185,43 +185,58 @@ public void printBacktrace(Node currentNode) {
...
-    public Backtrace getBacktrace(Node currentNode,
-                                  final int omit,
-                                  final boolean filterNullSourceSection,
-                                  DynamicObject exception,
-                                  Throwable javaThrowable) {
-        if (exception != null
-                && context.getOptions().BACKTRACES_OMIT_UNUSED
-                && DisablingBacktracesNode.areBacktracesDisabled()
-                && ModuleOperations.assignableTo(Layouts.BASIC_OBJECT.getLogicalClass(exception),
-                    context.getCoreLibrary().getStandardErrorClass())) {
```
Comment (eregon) [inline src/main/java/org/truffleruby/language/CallStackManager.java]: So the problem was that the exception was not passed in many cases, and we often don't have an exception instance at this point yet, so now we pass the exception class consistently to perform this check?

Comment (aardvark179) [reply src/main/java/org/truffleruby/language/CallStackManager.java]: There were several very distinct cases for filling in backtraces.
1. When raising an exception in ruby we have the exception and may need to fill in the stack trace
2. When raising an exception from Java we typically had not created the exception object yet.
3. Other stack related operations such as looking for callers, or creating a backtrace, etc.

I hoped to fix this with a fairly minimal change, but everywhere that created exceptions built its own backtrace, so I refactored things to reduce the number of places we did that and make it easier to sort things out.

Comment (eregon) [reply src/main/java/org/truffleruby/language/CallStackManager.java]: Thanks for the explanation!
